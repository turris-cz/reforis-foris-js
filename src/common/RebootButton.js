/*
 * Copyright (C) 2019-2024 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useState, useEffect } from "react";

import PropTypes from "prop-types";

import { useAPIPost } from "../api/hooks";
import { API_STATE } from "../api/utils";
import Button from "../bootstrap/Button";
import { Modal, ModalHeader, ModalBody, ModalFooter } from "../bootstrap/Modal";
import { useAlert } from "../context/alertContext/AlertContext";
import { ForisURLs } from "../utils/forisUrls";

function RebootButton(props) {
    const [triggered, setTriggered] = useState(false);
    const [modalShown, setModalShown] = useState(false);
    const [triggerRebootStatus, triggerReboot] = useAPIPost(ForisURLs.reboot);

    const [setAlert] = useAlert();
    useEffect(() => {
        if (triggerRebootStatus.state === API_STATE.ERROR) {
            setAlert(_("Reboot request failed."));
        }
    });

    const rebootHandler = () => {
        setTriggered(true);
        triggerReboot();
        setModalShown(false);
    };

    return (
        <>
            <RebootModal
                shown={modalShown}
                setShown={setModalShown}
                onReboot={rebootHandler}
            />
            <Button
                className="btn-danger"
                loading={triggered}
                disabled={triggered}
                onClick={() => setModalShown(true)}
                {...props}
            >
                {_("Reboot")}
            </Button>
        </>
    );
}

RebootModal.propTypes = {
    shown: PropTypes.bool.isRequired,
    setShown: PropTypes.func.isRequired,
    onReboot: PropTypes.func.isRequired,
};

function RebootModal({ shown, setShown, onReboot }) {
    return (
        <Modal shown={shown} setShown={setShown}>
            <ModalHeader setShown={setShown} title={_("Warning!")} />
            <ModalBody>
                <p>{_("Are you sure you want to restart the router?")}</p>
            </ModalBody>
            <ModalFooter>
                <Button onClick={() => setShown(false)}>{_("Cancel")}</Button>
                <Button className="btn-danger" onClick={onReboot}>
                    {_("Confirm reboot")}
                </Button>
            </ModalFooter>
        </Modal>
    );
}

export default RebootButton;
