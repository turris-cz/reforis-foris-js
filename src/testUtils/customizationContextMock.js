/*
 * Copyright (C) 2019-2022 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

window.CustomizationContext = React.createContext();

const deviceDetails = {
    kernel: "5.x.x",
    model: "Turris Omnia",
    os_branch: {
        mode: "branch",
        value: "hbs",
    },
    os_version: "6.x.x",
    reforis_version: "1.x.x",
    serial: 123456789,
};

const isCustomized = false;

function CustomizationContextMock({ children }) {
    return (
        <CustomizationContext.Provider value={{ deviceDetails, isCustomized }}>
            {children}
        </CustomizationContext.Provider>
    );
}

export { CustomizationContextMock };
