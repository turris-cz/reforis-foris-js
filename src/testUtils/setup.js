/*
 * Copyright (C) 2019-2024 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */
import React from "react";
import mockAxios from "jest-mock-axios";
import moment from "moment-timezone";
import "./mockGlobals";

// Setup axios cleanup
global.afterEach(() => {
    mockAxios.reset();
});

// Mock web sockets
window.WebSocket = jest.fn();

// Mock scrollIntoView
global.HTMLElement.prototype.scrollIntoView = () => {};

// Mock timezone utilities
jest.doMock("moment", () => {
    moment.tz.setDefault("UTC");
    return moment;
});
Date.now = jest.fn(() => new Date(Date.UTC(2019, 1, 1, 12, 13, 14)).valueOf());

// Mock Font Awesome v6 library
jest.mock("@fortawesome/react-fontawesome", () => ({
    FontAwesomeIcon: () => <i className="fa" />,
}));
