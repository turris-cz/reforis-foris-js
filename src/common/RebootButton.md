RebootButton component is a button that opens a modal dialog to confirm the
reboot of the device.

## Usage

```jsx
import React, { useEffect, createContext } from "react";
import RebootButton from "./RebootButton";
import { AlertContextProvider } from "../context/alertContext/AlertContext";

window.AlertContext = React.createContext();

const RebootButtonExample = () => {
    return (
        <AlertContextProvider>
            <div id="modal-container" />
            <div id="alert-container" />
            <RebootButton />
        </AlertContextProvider>
    );
};

<RebootButtonExample />;
```
