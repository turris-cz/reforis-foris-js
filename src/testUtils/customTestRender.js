/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

/* eslint import/export: "off" */

import React from "react";
import { UIDReset } from "react-uid";
import { StaticRouter } from "react-router-dom";
import { render } from "@testing-library/react";
import PropTypes from "prop-types";

import { AlertContextMock } from "./alertContextMock";
import { CustomizationContextMock } from "./customizationContextMock";

Wrapper.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node,
    ]),
};

function Wrapper({ children }) {
    return (
        <CustomizationContextMock>
            <AlertContextMock>
                <StaticRouter>
                    <UIDReset>{children}</UIDReset>
                </StaticRouter>
            </AlertContextMock>
        </CustomizationContextMock>
    );
}

const customTestRender = (ui, options) =>
    render(ui, { wrapper: Wrapper, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customTestRender as render };
