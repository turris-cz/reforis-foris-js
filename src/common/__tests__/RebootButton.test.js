/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import {
    fireEvent,
    getByText,
    queryByText,
    render,
    wait,
} from "customTestRender";
import mockAxios from "jest-mock-axios";
import { mockJSONError } from "testUtils/network";
import { mockSetAlert } from "testUtils/alertContextMock";

import RebootButton from "../RebootButton";

describe("<RebootButton/>", () => {
    let componentContainer;
    beforeEach(() => {
        const { container } = render(
            <>
                <div id="modal-container" />
                <RebootButton />
            </>
        );
        componentContainer = container;
    });

    it("Render.", () => {
        expect(componentContainer).toMatchSnapshot();
    });

    it("Render modal.", () => {
        expect(queryByText(componentContainer, "Confirm reboot")).toBeNull();
        fireEvent.click(getByText(componentContainer, "Reboot"));
        expect(componentContainer).toMatchSnapshot();
    });

    it("Confirm reboot.", () => {
        fireEvent.click(getByText(componentContainer, "Reboot"));
        fireEvent.click(getByText(componentContainer, "Confirm reboot"));
        expect(mockAxios.post).toHaveBeenCalledWith(
            "/reforis/api/reboot",
            undefined,
            expect.anything()
        );
    });

    it("Hold error.", async () => {
        fireEvent.click(getByText(componentContainer, "Reboot"));
        fireEvent.click(getByText(componentContainer, "Confirm reboot"));
        mockJSONError();
        await wait(() =>
            expect(mockSetAlert).toBeCalledWith("Reboot request failed.")
        );
    });
});
