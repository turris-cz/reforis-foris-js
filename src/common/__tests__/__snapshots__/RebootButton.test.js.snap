// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RebootButton/> Render modal. 1`] = `
<div>
  <div
    id="modal-container"
  >
    <div
      aria-labelledby="modal-title"
      aria-modal="true"
      class="modal fade show"
      role="dialog"
    >
      <div
        class="modal-dialog modal-dialog-centered"
        role="document"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
          >
            <h1
              class="modal-title fs-5"
            >
              Warning!
            </h1>
            <button
              aria-label="Close"
              class="btn-close"
              type="button"
            />
          </div>
          <div
            class="modal-body"
          >
            <p>
              Are you sure you want to restart the router?
            </p>
          </div>
          <div
            class="modal-footer"
          >
            <button
              class="btn btn-primary d-inline-flex justify-content-center align-items-center"
              type="button"
            >
              <span>
                Cancel
              </span>
            </button>
            <button
              class="btn btn-danger d-inline-flex justify-content-center align-items-center"
              type="button"
            >
              <span>
                Confirm reboot
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <button
    class="btn btn-danger d-inline-flex justify-content-center align-items-center"
    type="button"
  >
    <span>
      Reboot
    </span>
  </button>
</div>
`;

exports[`<RebootButton/> Render. 1`] = `
<div>
  <div
    id="modal-container"
  />
  <button
    class="btn btn-danger d-inline-flex justify-content-center align-items-center"
    type="button"
  >
    <span>
      Reboot
    </span>
  </button>
</div>
`;
